src/Shared/Dictionary/AdaptiveCapacityDictionary.cs

```diff
@@ -0,0 +1,752 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Internal.Dictionary
+{
+    /// <summary>
+    /// An <see cref="IDictionary{String, Object}"/> type to hold a small amount of items (4 or less in the common case).
+    /// </summary>
+    internal class AdaptiveCapacityDictionary<TKey, TValue> : IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue> where TKey : notnull
```

Comment:
Do envision using this where TKey is something other than a string in our code base? That would cut some of the size of this.

Reply:
This was something I was debating. We need to identify any place where we can use this to determine the key. I don't think it's a huge reduction though.